src/main/java/io/cdap/cdap/metrics/process/gcp/GoogleCloudMonitoringWriter.java

```diff
@@ -0,0 +1,219 @@
+  private static final String POLL_FREQUENCY_SECONDS = "poll.frequency.seconds";
+  private static final String MONITORING_ENDPOINT = "monitoring.endpoint";
+  private static final String CONFIG_FILE_PATH = "config.file.path";
+
+  private final AtomicLong lastEndTime = new AtomicLong(-1);
+  private Map<String, MetricsMapping> metricsMapping;
+  @Nullable
+  private MetricServiceClient metricServiceClient;
+  @Nullable
+  private String projectName;
+  private Map<String, String> autoFilledLabelMap;
+  private int pollFreqInSeconds;
+
+  @Override
+  public void write(Collection<MetricValues> metricValues) {
+    if (metricServiceClient == null) {
+      LOG.debug(
+        "MetricServiceClient is not available. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    if (projectName == null) {
+      LOG.debug(
+        "Project name is not configured. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    if (metricsMapping.isEmpty()) {
+      LOG.debug(
+        "Metrics are not mapped. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    LOG.debug("Writing metric values in GoogleCloudMonitoringWriter");
+    try {
+      Map<TimeSeriesMetadata, List<Long>> timeSeriesMap = createTimeSeriesMap(metricValues);
+      long endTime = System.currentTimeMillis() / 1000;
+      long startTime =
+        (lastEndTime.get() == -1 || lastEndTime.get() == endTime) ? endTime - pollFreqInSeconds
+          : lastEndTime.get();
+      List<TimeSeries> timeSeries = GoogleCloudMonitoringUtil.convertToMonitoringTimeSeries(timeSeriesMap,
+                                                                                            startTime, endTime);
+      LOG.debug(" Start time {}, end time {}", startTime, endTime);
+      GoogleCloudMonitoringUtil.send(timeSeries, metricServiceClient, projectName);
+      lastEndTime.getAndUpdate(value -> Math.max(value, endTime));
+    } catch (Exception exception) {
+      LOG.info("Exception while writing metric values in GoogleCloudMonitoringWriter.", exception);
+    }
+  }
+
+  @Override
+  public void initialize(MetricsWriterContext metricsWriterContext) {
+    MonitoringConfig config = getConfig(metricsWriterContext.getProperties().get(CONFIG_FILE_PATH));
+    this.metricsMapping = config.getMetricsMapping();
+    this.projectName = metricsWriterContext.getProperties().get(PROJECT);
+    this.pollFreqInSeconds = Integer.parseInt(metricsWriterContext.getProperties().get(POLL_FREQUENCY_SECONDS));
+    populateAutoFilledMap(metricsWriterContext.getProperties());
+
+    try {
+      String endPoint = metricsWriterContext.getProperties().get(MONITORING_ENDPOINT);
+      MetricServiceSettings metricServiceSettings =
+        MetricServiceSettings.newBuilder().setEndpoint(endPoint)
+          .build();
+      metricServiceClient = MetricServiceClient.create(metricServiceSettings);
```

Comment:
nit: can init be called multiple time? Shall we add defense code to protect against multi-invocation of init?

Reply:
Unlikely with the current caller, but added a check.

Reply:
nit: don't see updated the code.